base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,335 @@
+// should always correspond with the defines below, in sequence (except OPTION_INVALID)
+static PCWSTR OptionList[] = { L"?", L"nh" };
+
+#define OPTION_INVALID    -1
+#define OPTION_HELP       0
+#define OPTION_NOHEADER    1
+
+// the max string length PrintResString can handle
+#define RES_STR_MAXLEN 64
+
+// Print split line
+VOID PrintSplitLine(UINT Length)
+{
+    for (; Length; Length--)
+    {
+        ConPuts(StdOut, L"=");
+    }
+}
+
+// Print spaces
+VOID PrintSpace(UINT Length)
+{
+    ConPrintf(StdOut, L"%*ls", (INT)Length, L"");
+}
+
+// Print a string.
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+VOID PrintString(LPCWSTR String, UINT MaxWidth, BOOL bAlignLeft)
+{
+    ConPrintf(StdOut, bAlignLeft ? L"%-*.*ls" : L"%*.*ls", MaxWidth, MaxWidth, String);
+}
+
+// Print a string from resource
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+// The string WILL be truncated if it's longer than RES_STR_MAXLEN
+VOID PrintResString(HINSTANCE hInstance, UINT uID, UINT MaxWidth, BOOL bAlignLeft)
+{
+    if (!hInstance)
+        return;
+    
+    WCHAR StringBuffer[RES_STR_MAXLEN];
+    LoadStringW(hInstance, uID, StringBuffer, _countof(StringBuffer));
+    PrintString(StringBuffer, MaxWidth, bAlignLeft);
+}
+
+// Print a number, aligned to right.
+// MaxWidth is the width for printing.
+// the number WILL NOT be truncated if it's longer than MaxWidth
+VOID PrintNum(LONGLONG Number, UINT MaxWidth)
+{
+    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
+}
+
+// Print memory size using KB as unit, with comma-separated number, aligned to right.
+// MaxWidth is the width for printing.
+// the number WILL be truncated if it's longer than MaxWidth
+BOOL PrintMemory(SIZE_T MemorySizeByte, UINT MaxWidth, HINSTANCE hInstance)
+{
+    if (!hInstance)
+        return FALSE;
+
+    SIZE_T MemorySize = MemorySizeByte >> 10;
+
+    WCHAR NumberString[27] = { 0 }; // length 26 is enough to display ULLONG_MAX in decimal with comma, one more for zero-terminated.
+    C_ASSERT(sizeof(SIZE_T) <= 8);
+
+    PWCHAR pNumberStr = NumberString;
+
+    // calculate the length
+    UINT PrintLength = 0;
+    SIZE_T Tmp = MemorySize;
+    UINT Mod = 1;
+
+    do
+    {
+        Tmp /= 10;
+        PrintLength++;
+        Mod *= 10;
+    } while (Tmp);
+
+    for (UINT i = PrintLength; i; i--)
+    {
+        Mod /= 10;
+        *pNumberStr = L'0' + (MemorySize / Mod);
+        MemorySize %= Mod;
+        pNumberStr++;
+        
+        if (i != 1 && i % 3 == 1)
+        {
+            *pNumberStr = L',';
+            pNumberStr++;
+        }
+    }
+
+    WCHAR FormatStr[RES_STR_MAXLEN];
+    LoadStringW(hInstance, IDS_MEMORY_STR, FormatStr, _countof(FormatStr));
+
+    WCHAR String[RES_STR_MAXLEN + _countof(NumberString)];
+    
+    swprintf(String, FormatStr, NumberString);
```

Comment:
`swprintf` with no buffer overflow check is not secure.

Reply:
```
int swprintf_s(
   wchar_t *buffer,
   size_t sizeOfBuffer,
   const wchar_t *format,
   ...
);
```

Reply:
actually, I'm confused when I wrote this down yesterday.
this is the declaration of `swprint` I found in UCRT (and all other documents I can found)
```
int swprintf( wchar_t *buffer, size_t bufsz,
              const wchar_t* format, ... );
```

you can see the second param is `bufsz`, controlling maximum character could be written

but this is what I found in ReactOS: (at `reactos\sdk\include\crt\stdio.h`)
```
int __cdecl
    swprintf(
    _Out_ wchar_t*,
    _Printf_format_string_ const wchar_t*,
    ...);
```

the second param is missing, and I found another function pretty similar to `swprintf` I found elsewhere called `_swprintf_c`, which I failed to find any documents about it (possibly it's a ReactOS stuff?)

do you know the reason for this? 🤣 Is this a mistake when writing CRT for ReactOS, or does this having any purpose?

Reply:
> swprintf conforms to the ISO C Standard, which requires the second parameter, count, of type size_t. To force the old nonstandard behavior, define _CRT_NON_CONFORMING_SWPRINTFS. In a future version, the old behavior may be removed, so code should be changed to use the new conformant behavior.

(from MSDN)

Our code uses the old style (from WIn2003 and below), i.e. the so-called "nonstandard behaviour".

Reply:
I see...  then I guess using `swprintf_s`, just as what Hirofumi said, is better 🤣
seems this function doesn't have a diverge on standard/non-standard behaviour

Reply:
```txt
/base/applications/cmdutils/tasklist/tasklist.c:112: undefined reference to `swprintf_s'
```
Sorry, compilation is failing...

Reply:
> Sorry, compilation is failing... C11 is requirement.

Huh please NO. Just include the correct headers and it'll be fixed.

Reply:
You can use as well _snwprintf() (I would say I would prefer that one since it remains CRT function).

Reply:
actually, I think it doesn't matter whether if the function is in CRT or not, we've already used tons of WinAPI even NT API in this program 🤣